packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts

```diff
@@ -925,6 +1028,93 @@ export const useAzureCommunicationCallAdapter = (
   return adapter;
 };
 
+/**
+ * A custom React hook to simplify the creation of {@link CallAdapter}.
+ *
+ * Similar to {@link createAzureCommunicationCallAdapter}, but takes care of asynchronous
+ * creation of the adapter internally.
+ *
+ * Allows arguments to be undefined so that you can respect the rule-of-hooks and pass in arguments
+ * as they are created. The adapter is only created when all arguments are defined.
+ *
+ * Note that you must memoize the arguments to avoid recreating adapter on each render.
+ * See storybook for typical usage examples.
+ *
+ * @public
+ */
+export const useAzureCommunicationCallAdapter = (
+  /**
+   * Arguments to be passed to {@link createAzureCommunicationCallAdapter}.
+   *
+   * Allows arguments to be undefined so that you can respect the rule-of-hooks and pass in arguments
+   * as they are created. The adapter is only created when all arguments are defined.
+   */
+  args: Partial<AzureCommunicationCallAdapterArgs>,
+  /**
+   * Optional callback to modify the adapter once it is created.
+   *
+   * If set, must return the modified adapter.
+   */
+  afterCreate?: (adapter: CallAdapter) => Promise<CallAdapterCommon>,
+  /**
+   * Optional callback called before the adapter is disposed.
+   *
+   * This is useful for clean up tasks, e.g., leaving any ongoing calls.
+   */
+  beforeDispose?: (adapter: CallAdapter) => Promise<void>
+): CallAdapter | undefined => {
+  return useAzureCommunicationCallAdapterGeneric(
+    args,
+    afterCreate as (adapter: CallAdapter) => Promise<CallAdapter>,
```

Comment:
This downcast is not safe. What if Contoso provides a callback that actually returns only a `Promise<CallAdapterCommon>`?

That said, I think we can keep the signature `afterCreate?: (adapter: CallAdapter) => Promise<CallAdapter>` because before this change, Contoso _must have_ used a callback with that signature (this PR only splits the definition of `CallAdapter` into `CallAdapterCommon` and `CallAdapter`) so that this downcast won't be necessary.

Same comment below.

Reply:
That's a point I wanted to discuss - lemme resolve all the other comments from Internal review and this PR, then we can chat about this a bit later
It is just because I found myself had a really hard time to implement a common function in contoso side